// @google-cloud/dialogflow-cx/build/src/v3beta1/index.js
"use strict";
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhooksClient = exports.VersionsClient = exports.TransitionRouteGroupsClient = exports.TestCasesClient = exports.SessionsClient = exports.SessionEntityTypesClient = exports.SecuritySettingsServiceClient = exports.PagesClient = exports.IntentsClient = exports.FlowsClient = exports.ExperimentsClient = exports.EnvironmentsClient = exports.EntityTypesClient = exports.DeploymentsClient = exports.ChangelogsClient = exports.AgentsClient = void 0;
var agents_client_1 = require("./agents_client");
Object.defineProperty(exports, "AgentsClient", { enumerable: true, get: function () { return agents_client_1.AgentsClient; } });
var changelogs_client_1 = require("./changelogs_client");
Object.defineProperty(exports, "ChangelogsClient", { enumerable: true, get: function () { return changelogs_client_1.ChangelogsClient; } });
var deployments_client_1 = require("./deployments_client");
Object.defineProperty(exports, "DeploymentsClient", { enumerable: true, get: function () { return deployments_client_1.DeploymentsClient; } });
var entity_types_client_1 = require("./entity_types_client");
Object.defineProperty(exports, "EntityTypesClient", { enumerable: true, get: function () { return entity_types_client_1.EntityTypesClient; } });
var environments_client_1 = require("./environments_client");
Object.defineProperty(exports, "EnvironmentsClient", { enumerable: true, get: function () { return environments_client_1.EnvironmentsClient; } });
var experiments_client_1 = require("./experiments_client");
Object.defineProperty(exports, "ExperimentsClient", { enumerable: true, get: function () { return experiments_client_1.ExperimentsClient; } });
var flows_client_1 = require("./flows_client");
Object.defineProperty(exports, "FlowsClient", { enumerable: true, get: function () { return flows_client_1.FlowsClient; } });
var intents_client_1 = require("./intents_client");
Object.defineProperty(exports, "IntentsClient", { enumerable: true, get: function () { return intents_client_1.IntentsClient; } });
var pages_client_1 = require("./pages_client");
Object.defineProperty(exports, "PagesClient", { enumerable: true, get: function () { return pages_client_1.PagesClient; } });
var security_settings_service_client_1 = require("./security_settings_service_client");
Object.defineProperty(exports, "SecuritySettingsServiceClient", { enumerable: true, get: function () { return security_settings_service_client_1.SecuritySettingsServiceClient; } });
var session_entity_types_client_1 = require("./session_entity_types_client");
Object.defineProperty(exports, "SessionEntityTypesClient", { enumerable: true, get: function () { return session_entity_types_client_1.SessionEntityTypesClient; } });
var sessions_client_1 = require("./sessions_client");
Object.defineProperty(exports, "SessionsClient", { enumerable: true, get: function () { return sessions_client_1.SessionsClient; } });
var test_cases_client_1 = require("./test_cases_client");
Object.defineProperty(exports, "TestCasesClient", { enumerable: true, get: function () { return test_cases_client_1.TestCasesClient; } });
var transition_route_groups_client_1 = require("./transition_route_groups_client");
Object.defineProperty(exports, "TransitionRouteGroupsClient", { enumerable: true, get: function () { return transition_route_groups_client_1.TransitionRouteGroupsClient; } });
var versions_client_1 = require("./versions_client");
Object.defineProperty(exports, "VersionsClient", { enumerable: true, get: function () { return versions_client_1.VersionsClient; } });
var webhooks_client_1 = require("./webhooks_client");
Object.defineProperty(exports, "WebhooksClient", { enumerable: true, get: function () { return webhooks_client_1.WebhooksClient; } });
